README.Rmd

```diff
@@ -18,37 +18,274 @@ knitr::opts_chunk$set(
+)
+```
+
+# Getting started
 
 ## Installation
 
 You can install the development version of cfazuR from [GitHub](https://github.com/) with:
 
-``` r
-# install.packages("pak")
+```{r eval=FALSE}
 pak::pak("CDCgov/cfazuR")
 ```
 
-## Example
+## Credential setup (required)
+
+To connect to Azure storage resources with cfazuR, you will need to be able to
+authenticate (i.e. log in) to Azure from R. You also need to look up certain
+Azure credentials (i.e. account information) and store them in a location that
+is secure, but accessible to R.
+
+Follow these the steps carefully the first time you set up the package, and
+refer back to this page if you ever need to update your credentials.
 
-This is a basic example which shows you how to solve a common problem:
+### 1. Learn how to keep your credentials secure
 
-```{r example}
-library(cfazuR)
-## basic example code
+**Big picture: Treat your credentials like passwords.** Do not commit them to GitHub, post
+them on the internet, send them in unencrypted emails, etc. This is especially
+important for service principals, which are essentially passwords. But it's a
+good idea to follow these security guidelines for all of your credentials, the
+same way you're probably already used to doing for other kinds of usernames and
+account numbers.
+
+**1.1 [Install and activate detect-secrets](https://potential-adventure-637em36.pages.github.io/docs/pre_commit.html#getting-started-installing-pre-commit)**
+as a pre-commit hook. This will check
+your code for secrets before you commit to GitHub and raise an error if it
+detects a problem.
+
+**1.2 Never hard-code your credentials**
+
+```{r eval=FALSE}
+## WHAT NOT TO DO:
+az_login(
+  tenant = "1234-abcd-0987-zyxw",
+  app = "my-actual-client-id",
+  password = "my-actual-service-principal" # pragma: allowlist secret
+)
 ```
 
+If you accidentally push this code to GitHub, your secrets are now on the
+internet. Instead, write code that represents your credentials as variables
+without ever showing the underlying values:
+```{r eval=FALSE}
+## DO THIS INSTEAD
+##  Load credentials stored in a separate, .json, .yaml, or .toml file
+cred_list <- read_json(my_credentials.json)
+## or
+## Store credentials as environment variables and load them at runtime
+cred_list <- cfazuR::get_env_creds() # Load env vars
+az_login(
+  tenant = cred_list$az_tenant_id,
+  app = cred_list$az_client_id,
+  password = cred_list$az_service_principal
+)
+```
+
+### 2. Look up your Azure Credentials
+
+To log in using [client-based authentication]() you will need:
+
+1. Client ID
+2. Tenant ID
+3. Service principal
+
+To find these, log in to the Azure portal (https://portal.azure.com/), then
+search for "app registrations" and click on the relevant icon. If an app
+registration exists, click on it to reveal a page showing the associated
+client ID, and tenant ID. From this page, click again below "Certificates and
+secrets" to reveal the service principal value.
```

Comment:
I'm having trouble following these instructions. I found the page but I'm unclear whether "If an app
registration exists" means I should make a new app registration or if I should look for something in particular.

Reply:
SO I clicked on "App registrations", which then gave me three options

- All applications. This listed out a bunch of things. For instance, `EDAV-CFA-Predict-NNHT-SP`. I clicked on that one, went to the "Certificates & secrets". From there, I was able to see the first three characters of the Value, but not the rest. I was able to copy the Secret ID, which I *think* can be used via the `az` command line to get the value of the secret?
- Owned applications. Nothing here.
- Deleted applications. Nothing here.

Reply:
I tried `az keyvault secret show --id <the uuid of the NNH teams's SP>`, but it told me it was not a valid ID